Archive/Imo/Imo1997Q3.lean

```diff
@@ -0,0 +1,129 @@
+/-
+Copyright (c) 2025 Jeremy Tan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jeremy Tan
+-/
+import Mathlib.Algebra.BigOperators.Ring.Finset
+import Mathlib.Algebra.Group.Submonoid.Membership
+import Mathlib.Data.Real.Basic
+import Mathlib.Data.Sign.Defs
+import Mathlib.GroupTheory.Perm.Sign
+
+/-!
+# IMO 1997 Q3
+
+Let $x_1, x_2, \dots, x_n$ be real numbers satisfying the conditions
+$|x_1 + x_2 + \cdots + x_n| = 1$ and $|x_i| ≤ \frac{n+1}2$ for $i = 1, 2, \dots, n$.
+Show that there exists a permutation $y_1, y_2, \dots, y_n$ of $x_1, x_2, \dots, x_n$ such that
+$|y_1 + 2y_2 + \cdots + ny_n| ≤ \frac{n+1}2$.
+
+# Solution
+
+For a permutation $π$ let $S(π) = \sum_{i=1}^n i x_{π(i)}$. We wish to show that there exists $π$
+with $|S(π)| ≤ \frac{n+1}2$.
+
+Suppose the contrary, that all permutations satisfy $\frac{n+1}2 < |S(π)|$. Then we note that for
+two permutations $π, π'$ differing by a swap of adjacent elements, say $x_k$ and $x_{k+1}$,
+$|S(π) - S(π')| = |x_k - x_{k+1}| ≤ n + 1$. This being the size of the interval
+$[-\frac{n+1}/2, \frac{n+1}/2]$ forbidden to the $|S(π)|$, $S(π)$ and $S(π')$ must be on the
+same side of said interval, i.e. they have the same sign.
+By induction all the $S(\pi)$ have the same sign.
+
+But now consider the identity permutation $1$ and the reverse permutation $R$. We have
+$|S(1) + S(R)| = |(1 + n)x_1 + \cdots + (n + 1)x_n| = (n + 1)|x_1 + \cdots + x_n| = n + 1$.
+Since $S(1)$ and $S(R)$ have the same sign yet are strictly larger than $\frac{n+1}2$ in
+absolute value, $|S(1) + S(R)| > n + 1$, which yields a contradiction. Therefore the initial
+assumption that all permutations satisfy $\frac{n+1}2 < |S(π)|$ must be false; the result follows.
+-/
+
+namespace Imo1997Q3
+
+open Equiv Fin Finset SignType
+
+variable {n : ℕ}
+
+/-- The sum $x_1 + 2x_2 + \cdots + nx_n$ mentioned in the problem. -/
+def S (x : Fin n → ℝ) (p : Perm (Fin n)) : ℝ :=
+  ∑ i, (i + 1) * x (p i)
+
+lemma sign_eq_of_abs_sub_le {a b c : ℝ} (ha : c / 2 < |a|) (hb : c / 2 < |b|) (hc : 0 < c)
+    (hs : |a - b| ≤ c) : sign a = sign b := by
+  rw [lt_abs] at ha hb
+  rcases ha with ha | ha
+  · rw [sign_pos ((half_pos hc).trans ha)]
+    rcases hb with hb | hb
+    · rw [sign_pos ((half_pos hc).trans hb)]
+    · have m := add_lt_add ha hb; rw [add_halves, ← sub_eq_add_neg] at m
+      exact absurd (le_of_abs_le hs) (not_le.mpr m)
+  · rw [sign_neg (neg_pos.mp <| (half_pos hc).trans ha)]
+    rcases hb with hb | hb
+    · have m := add_lt_add ha hb; rw [add_halves, ← sub_eq_neg_add, ← neg_sub, lt_neg] at m
+      exact absurd (neg_le_of_abs_le hs) (not_le.mpr m)
+    · rw [sign_neg (neg_pos.mp <| (half_pos hc).trans hb)]
```

Comment:
Since there's nothing interesting mathematically going on here, I'd be inclined just to brute-force it (and the next lemma can probably be brute-forced similarly).

```lean
lemma sign_eq_of_abs_sub_le {a b c : ℝ} (ha : c / 2 < |a|) (hb : c / 2 < |b|) (hc : 0 < c)
    (hs : |a - b| ≤ c) : sign a = sign b := by
  rcases lt_trichotomy 0 a with ha' | rfl | ha' <;> rcases lt_trichotomy 0 b with hb' | rfl | hb' <;> simp_all [abs_of_pos, abs_of_neg, abs_le] <;> linarith
```

(Quite possibly there are better / shorter ways of doing this with automation.)

Reply:
Done.